### ISSUE TITLE
Command line arguments not passing to my CLI tool

### ISSUE THREAD

**Issue by mwpowellhtx** [dotnet/cli#11598]
Possibly also stuck in a ``dotnet`` cache somewhere.

## Steps to reproduce

```Xml
<Exec Command="dotnet mytool -- --version" ConsoleToMSBuild="true" StandardOutputImportance="normal" ContinueOnError="true">
  <Output TaskParameter="ConsoleOutput" PropertyName="_MyToolVersion" />
</Exec>
```

Yes, I do have a ``dotnet-mytool.dll`` registered:

```Xml
<ItemGroup>
  <DotNetCliToolReference Include="dotnet-mytool" Version="*" />
</ItemGroup>
```

However, version(s) of my tool appear to be somehow cached, because I am not finding behavior consistent with my troubleshooting efforts. In the CLI tool itself:

```C#
using static String;
// ...
public static TextWriter Out { get; internal set; } = Console.Out;

public static TextWriter Error { get; internal set; } = Console.Error;

public static int Main(string[] args)
{
    Out.WriteLine($@"The arguments were: `{Join(" ", args)}´.");
    Error.WriteLine($@"The arguments were: `{Join(" ", args)}´.");
    return 42;
}
```

Yet, when I am trying to integration test, I am finding the message ``The output directory must be specified.``, which is a previous version of the CLI tool, and would be consistent with not having been given the expected command line arguments.

Now, I did look at file properties, and these are both the same for the actual project output as well as the project dependency. So I am 99.99% confident that the CLI tool is landing as it should, notwithstanding the fact I cannot square the apparent caching issue.

## Expected  behavior
I have integration tests in which I can directly invoke the ``Program.Main``. This works as expected. This is as robust an integration test as I can imagine short of invoking the CLI tool at the command line, never mind via *Microsoft Build Target*, which is where I ultimately want it to work correctly.

## Actual behavior
For whatever reason, the command line arguments do not appear to be relayed to my CLI tool. I am fairly confident, however, that my CLI plumbing is at least correct because I am finding error messages reported that would indicate at least ``dotnet mytool -- ...`` did get invoked. Without the expected command line arguments. Or at least, the error is obvious when the plumbing is incorrect.

## Environment data
`dotnet --info` attached.

[dotnet-info.txt](https://github.com/dotnet/cli/files/3312275/dotnet-info.txt)

**Comment by mwpowellhtx** [comment]
Another thought occurred to me... ``The output directory must be specified.`` could actually be a *dotnet CLI* message itself, although I cannot confirm that. I do happen to have an internal message, but it is, ``An output directory must be specified.``. So I think this is more of a *dotnet CLI* message, possibly an *Microsoft Build* message concerning the ``<Exec/>``.

**Comment by mwpowellhtx** [comment]
Possibly identified an issue, ``prefercliruntime`` does not travel with the CLI tool when a project depends on the CLI tool project. I'll see about working around that.

**Comment by mwpowellhtx** [comment]
I discovered that ``msbuild``, ``dotnet``, ``nuget``, not necessarily in that order, or one of its surrogates, apparently `discovered´, so to speak, a CLI tool nuget package and tried to install it in my ``packages\.tools`` cache. While accurate, there is a naming conflict, I am not sure why the parts involved would not fall back on the project level artifacts that are right there for it to use. This at least explains the messages, and especially the version number. But it does not explain how I can persuade one of the surrogates to do the right thing and `just work´ with my project level artifacts.

**Comment by livarcocc** [comment]
@mwpowellhtx Can you summarize what you are trying to do and how? Are you producing a global tool and installing it and then trying to invoke it through MSBuild?

**Comment by mwpowellhtx** [comment]
@livarcocc I'm not sure what you mean by ´global tool´. I am developing a CLI tool, yes. Invoked as ``dotnet mytool -- <mytool-args/>``, and so on; which means it will get published as ``dotnet-mytool.dll`` asset. At least that's how I understand the build pipeline. To facilitate that, I am using ``<ProjectReference/>`` in order to ensure that the tool, and, ostensibly, any package referenced assets it requires, is available in for use when I need it.

**Comment by mwpowellhtx** [comment]
@livarcocc Any ideas? Bare minimum, I think my my courses of action must include renaming my target CLI tool. That much is clear. But I would still like to get a better understanding as to how I might better persuade ``dotnet`` or any of its downstream surrogates to ´just work´ and do the right thing.

**Comment by mwpowellhtx** [comment]
@livarcocc Probably separate issue, but corollary to this is my CLI tool references are not traveling with the project when I make a ``<ProjectReference/>``. Perhaps that could also be clarified; is there something about CLI tools that prohibits them from carrying package references with them?

**Comment by mwpowellhtx** [comment]
@livarcocc The bottom line for me, I think, is how these are rolled up in the deliverable build targets. Here are a few examples:

```Xml
<Exec Command="dotnet $(MyToolPath) --version" ConsoleToMsBuild="true" StandardOutputImportance="normal" ContinueOnError="true">
  <Output TaskParameter="ConsoleOutput" PropertyName="_MyToolVersion"/>
</Exec>
```

And:

```Xml
<Exec Command="dotnet $(MyToolPath) --response &quot;$(_MyToolResponseFilePath)&quot;" StandardOutputImportance="normal"  ContinueOnError="true" />
```

Although another example I saw was this, which I'm not sure whether *dotnet* is taking the response file and passing the line separated arguments to *dotnet-mytool.dll*?

```Xml
<Exec Command="dotnet $(MyToolPath) &quot;%40$(_MyToolResponseFilePath)&quot;" StandardOutputImportance="normal"  ContinueOnError="true" />
```

Which has a couple of questions there for me:

1. How to pass arguments to my assets.
1. As I understand it, ``MyToolPath`` can be the tool name, or a path to the tool DLL. However, when I tried this, I was finding *dotnet* wanting a tool named something like this: ``dotnet-path/to/mytool/dotnet-mytool.dll``. Clearly incorrect.
1. And in a way that respects my build pipeline, does not jump out to surrogate, i.e. *nuget*, processes to get packages I do not need, because those assets are part of my build pipeline, albeit with a plausible naming conflict.

**Comment by mwpowellhtx** [comment]
I think I've potentially found the issue, but I need to troubleshoot that a bit.

I had this:

```Xml
<MyToolPath>bin\$(Configuration)\$(TargetFramework)\dotnet-mytool</MyToolPath>
```

Assuming the directory was correct, *dotnet-mytool* clearly was not. I think it may in fact need to be this:

```Xml
<MyToolPath>bin\$(Configuration)\$(TargetFramework)\dotnet-mytool.dll</MyToolPath>
```

Including the file extension.

**Comment by mwpowellhtx** [comment]
I provided the full path, including extension, to my targets, and got the expected response:

```
The arguments were: `--response obj\Debug\netcoreapp2.1\MyTool.csproj.rsp'.
```

My verification ``Program.Main``, is:

```C#
public static int Main(string[] args)
{
    Out.WriteLine($@"The arguments were: `{string.Join(" ", args)}´.");
    Error.WriteLine($@"The arguments were: `{string.Join(" ", args)}´.");
    return 42;
}
```

So, what's left for me is to reconnect the actual tooling code and see what happens.

**Comment by mwpowellhtx** [comment]
@livarcocc Okay, I am confused. It seemed to work one time, perhaps an initial VS session build attempt? I am receiving an exception during the build. But the odd part is, the ``Program.Main`` does get invoked?

```
6>Preparing to run my tool using intermediate directory `obj\Debug\netcoreapp2.1\´ ...
6>Evaluating `dotnet -d ..\MyTool\bin\Debug\netcoreapp2.1\dotnet-mytool.dll --version´ ...
6>Microsoft.DotNet.Cli.Utils.CommandUnknownException: No executable found matching command "dotnet-..\MyTool\bin\Debug\netcoreapp2.1\dotnet-mytool.dll"
6>   at Microsoft.DotNet.Cli.Utils.Command.Create(ICommandResolverPolicy commandResolverPolicy, String commandName, IEnumerable`1 args, NuGetFramework framework, String configuration, String outputPath, String applicationName)
6>   at Microsoft.DotNet.Cli.Utils.Command.Create(String commandName, IEnumerable`1 args, NuGetFramework framework, String configuration, String outputPath, String applicationName)
6>   at Microsoft.DotNet.Cli.Program.ProcessArgs(String[] args, ITelemetry telemetryClient)
6>   at Microsoft.DotNet.Cli.Program.Main(String[] args)
6>G:\Source\Spikes\MyTool\Prototype\src\MyTool.Tasks\build\MyTool.BuildTime.targets(81,5): warning MSB3073: The command "dotnet -d ..\MyTool\bin\Debug\netcoreapp2.1\dotnet-mytool.dll --version" exited with code 1.
6>Running MyTool vTelemetry is: Enabled;projectfactory: MSBUILD_EXE_PATH = C:\Program Files\dotnet\sdk\2.1.507\MSBuild.dll;projectfactory: MSBuild project path = G:\Source\Spikes\MyTool.Core\Prototype\src\MyTool.Integration.Tests\MyTool.Integration.Tests.csproj;projecttoolscommandresolver: resolving commandspec from 1 Tool Libraries.;projecttoolscommandresolver: Attempting to resolve command spec from tool dotnet-mytool;projecttoolscommandresolver: nuget packages root:;- G:\Source\Spikes\MyTool\Prototype\src\packages;- C:\Program Files\dotnet\sdk\NuGetFallbackFolder;projecttoolscommandresolver: found tool lockfile at : G:\Source\Spikes\MyTool\Prototype\src\packages\.tools\dotnet-mytool\0.6.1\netcoreapp2.1\project.assets.json;projecttoolscommandresolver: expect deps.json at: G:\Source\Spikes\MyTool\Prototype\src\packages\.tools\dotnet-mytool\0.6.1\netcoreapp2.1\dotnet-codegen.deps.json;projecttoolscommandresolver: attempting to create commandspec;packagedcommandspecfactory: attempting to find command dotnet-..\MyTool\bin\Debug\netcoreapp2.1\dotnet-mytool.dll in dotnet-mytool;packagedcommandspecfactory: failed to find toolAssembly for dotnet-..\MyTool\bin\Debug\netcoreapp2.1\dotnet-mytool.dll;projecttoolscommandresolver: commandSpec is null.;projecttoolscommandresolver: failed to resolve commandspec from library.;Microsoft.DotNet.Cli.Utils.CommandUnknownException: No executable found matching command "dotnet-..\MyTool\bin\Debug\netcoreapp2.1\dotnet-mytool.dll";at Microsoft.DotNet.Cli.Utils.Command.Create(ICommandResolverPolicy commandResolverPolicy, String commandName, IEnumerable`1 args, NuGetFramework framework, String configuration, String outputPath, String applicationName);at Microsoft.DotNet.Cli.Utils.Command.Create(String commandName, IEnumerable`1 args, NuGetFramework framework, String configuration, String outputPath, String applicationName);at Microsoft.DotNet.Cli.Program.ProcessArgs(String[] args, ITelemetry telemetryClient);at Microsoft.DotNet.Cli.Program.Main(String[] args) ...
6>The arguments were: `--response obj\Debug\netcoreapp2.1\MyTool.Integration.Tests.csproj.rsp'.
```

I do have telemetry engaged to try and further decipher just what the heck is going on here. It seems to me that possibly ``dotnet`` is trying to farm out to the NuGet Package source for the "conflicting" NuGet package. However, the odd part is, pleasantly so, at least I am finding the expected output to follow. I just wonder how I can persuade ``dotnet`` not to defer to *NuGet* package sources unless it absolutely must.

With ``dotnet -d`` enabled, maybe this is just the detailed log reporting every step, I do not know.

Thoughts?

**Comment by mwpowellhtx** [comment]
@livarcocc Did a bit more troubleshooting. I am pretty confident now that messages are making it to my tooling properly and it is just a matter of solving the couple of issues I've uncovered to do with the tooling itself. Thanks for sounding off concerning any ``dotnet`` bits, much appreciated.

**Comment by mwpowellhtx** [comment]
@livarcocc Just wanted to circle around and let you know. I had a minor breakthrough the other day as I mentioned, ``dotnet`` was landing on ``mytool`` tooling successfully. Followed closely by a successful end-to-end integration. I have a bit of unit testing to do before I can call it done-done, but I am happy with that major breakthrough. Thanks again.
